File: problem2.java
// Time Complexity : O(M * N)
// Space Complexity : O(M * N) - Memoization table
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no

// the minimum falling path sum across all possible starting columns in the first row.
class problem2 {
    public int minFallingPathSum(int[][] matrix) {
        int minFallingSum = Integer.MAX_VALUE;
        Integer memo[][] = new Integer[matrix.length][matrix[0].length];

        for (int startCol = 0; startCol < matrix.length; startCol++) {
            minFallingSum = Math.min(minFallingSum,
                    findMinFallingPathSum(matrix, 0, startCol, memo));
        }
        return minFallingSum;
    }

    public int findMinFallingPathSum(int[][] matrix, int row, int col, Integer[][] memo) {
        if (col < 0 || col == matrix.length) {
            return Integer.MAX_VALUE;
        }
        if (row == matrix.length - 1) {
            return matrix[row][col];
        }
        if (memo[row][col] != null) {
            return memo[row][col];
        }

        int left = findMinFallingPathSum(matrix, row + 1, col, memo);
        int middle = findMinFallingPathSum(matrix, row + 1, col + 1, memo);
        int right = findMinFallingPathSum(matrix, row + 1, col - 1, memo);

        memo[row][col] = Math.min(left, Math.min(middle, right)) + matrix[row][col];
        return memo[row][col];
    }
}

File: Problem2.cs
public class LeastFallingPathSum
	{
        // Time Complexity : O(n^2) 
        // Space Complexity : O(1)
        // Did this code successfully run on Leetcode : Yes
        // Any problem you faced while coding this : No
        public int MinFallingPathSum(int[][] matrix)
        {
            if (matrix == null || matrix.Length == 0)
                return 0;

            int n = matrix.Length;
            int minSum = Int32.MaxValue;

            for(int i = n - 2; i >=0; i--)
            {
                for(int j = 0; j < n; j++)
                {
                    if (j == 0)
                    {
                        matrix[i][j] = matrix[i][j] + Math.Min(matrix[i + 1][j], matrix[i + 1][j + 1]);
                    }
                    else if (j == n - 1)
                    {
                        matrix[i][j] = matrix[i][j] + Math.Min(matrix[i + 1][j], matrix[i + 1][j - 1]);
                    }
                    else
                    {
                        matrix[i][j] = matrix[i][j] + Math.Min(matrix[i + 1][j - 1],Math.Min(matrix[i + 1][j], matrix[i + 1][j + 1]));
                    }
                }
            }
            for(int j = 0; j < n; j++)
            {
                minSum = Math.Min(minSum, matrix[0][j]);
            }

            return minSum;
        }
}
